src/number_theory/legendre_symbol/quadratic_char.lean

```diff
@@ -0,0 +1,339 @@
+by simp only [quadratic_char, eq_self_iff_true, if_true, id.def]
+
+@[simp]
+lemma quadratic_char_one : quadratic_char F 1 = 1 :=
+by simp only [quadratic_char, one_ne_zero, is_square_one, if_true, if_false, id.def]
+
+/-- For nonzero `a : F`, `quadratic_char F a = 1 ↔ is_square a`. -/
+lemma quadratic_char_one_iff_is_square {a : F} (ha : a ≠ 0) :
+  quadratic_char F a = 1 ↔ is_square a :=
+by { simp [quadratic_char, ha, (dec_trivial : (-1 : ℤ) ≠ 1)], tauto }
+
+/-- The quadratic character takes the value `1` on nonzero squares. -/
+lemma quadratic_char_sq_one' {a : F} (ha : a ≠ 0) : quadratic_char F (a ^ 2) = 1 :=
+by simp only [quadratic_char, ha, pow_eq_zero_iff, nat.succ_pos', is_square_sq, if_true, if_false]
+
+/-- If `ring_char F = 2`, then `quadratic_char F` takes the value `1` on nonzero elements. -/
+lemma quadratic_char_eq_one_of_char_two (hF : ring_char F = 2) {a : F} (ha : a ≠ 0) :
+  quadratic_char F a = 1 :=
+begin
+  simp [quadratic_char, ha],
+  intro h,
+  exfalso,
+  exact h (finite_field.is_square_of_char_two hF a),
+end
+
+/-- If `ring_char F` is odd, then `quadratic_char F a` can be computed in
+terms of `a ^ (fintype.card F / 2)`. -/
+lemma quadratic_char_eq_pow_of_char_ne_two (hF : ring_char F ≠ 2) {a : F} (ha : a ≠ 0) :
+  quadratic_char F a = if a ^ (fintype.card F / 2) = 1 then 1 else -1 :=
+begin
+  simp only [quadratic_char, ha, if_false],
+  simp_rw finite_field.is_square_iff hF ha,
+end
+
+/-- The quadratic character is multiplicative. -/
+lemma quadratic_char_mul (a b : F) :
+  quadratic_char F (a * b) = quadratic_char F a * quadratic_char F b :=
+begin
+  by_cases ha : a = 0,
+  { rw [ha, zero_mul, quadratic_char_zero, zero_mul], },
+  -- now `a ≠ 0`
+  by_cases hb : b = 0,
+  { rw [hb, mul_zero, quadratic_char_zero, mul_zero], },
+  -- now `a ≠ 0` and `b ≠ 0`
+  have hab := mul_ne_zero ha hb,
+  by_cases hF : ring_char F = 2,
+  { -- case `ring_char F = 2`
+    rw [quadratic_char_eq_one_of_char_two hF ha,
+        quadratic_char_eq_one_of_char_two hF hb,
+        quadratic_char_eq_one_of_char_two hF hab,
+        mul_one], },
+  { -- case of odd characteristic
+    rw [quadratic_char_eq_pow_of_char_ne_two hF ha,
+        quadratic_char_eq_pow_of_char_ne_two hF hb,
+        quadratic_char_eq_pow_of_char_ne_two hF hab,
+        mul_pow],
+    cases finite_field.pow_dichotomy hF hb with hb' hb',
+    { simp only [hb', mul_one, eq_self_iff_true, if_true], },
+    { have h := finite_field.neg_one_ne_one_of_char_ne_two hF, -- `-1 ≠ 1`
+      simp only [hb', h, mul_neg, mul_one, if_false, ite_mul, neg_mul],
+      cases finite_field.pow_dichotomy hF ha with ha' ha';
+        simp only [ha', h, neg_neg, eq_self_iff_true, if_true, if_false], }, },
+end
+
+/-- The quadratic character is a homomorphism of monoids with zero. -/
+def quadratic_char_hom : F →*₀ ℤ :=
+{ to_fun := quadratic_char F, map_zero' := quadratic_char_zero, map_one' := quadratic_char_one,
+  map_mul' := quadratic_char_mul }
```

Comment:
```suggestion
/-- The quadratic character is a homomorphism of monoids with zero. -/
@[simps] def quadratic_char_hom : F →*₀ ℤ :=
{ to_fun := quadratic_char F,
  map_zero' := quadratic_char_zero,
  map_one' := quadratic_char_one,
  map_mul' := quadratic_char_mul }
```

Reply:
FYI: `@[simps]` auto-generates a simp-lemma that says that `quadratic_char_hom F a` unfolds to `quadratic_char F a`.